test/helpers/navigate.js

```diff
@@ -0,0 +1,12 @@
+/**
+ * Navigate to a specific hashtag route in the page, and wait for docsify to handle navigation.
+ * @param {import('playwright-core').Page} page
+ * @param {string} route
+ */
+async function navigateToRoute(page, route) {
+  await page.evaluate(r => (window.location.hash = r), route);
+  const mainElm = await page.waitForSelector('#main');
+  await mainElm.waitForElementState('stable');
+}
+
+module.exports.navigateToRoute = navigateToRoute;
```

Comment:
First of all, you are correct, there's no need to await `#main` to be stable. I'll remove that.

The reason I did not use `page.goto` is that our test server does not support appending hashes to it, and returns 404. Instead, I opted to use `page.evaluate(() => (location.hash = ...))`. `navigateToRoute` is just a simple wrapper around that, meant to serve as a meaningful alias conveying the action's intent.

Should I update it with an optional third parameter that tells which type of routing we use? e.g.:
```js
async function navigateToRoute(page, route, { routerMode = 'hash' } = {}) {
  if (routerMode === 'hash') {
    await page.evalulate(() => location.hash = ... );
  } else {
    await page.evalulate(() => history.pushState(...));
  }
}
```

wdyt?

Reply:
We can get the `routerMode` via the DOM instead of requiring devs to specify it as an argument.

Also, after looking at the `page.goto()` [options](https://playwright.dev/docs/api/class-page#page-goto) more carefully, I realized most of them wouldn't apply because we aren't actually navigating to a new page. The most interesting option allows us to wait for `networkidle` which we can do via `page.waitForLoadState()`:

```js
async function navigateToRoute(page, route) {
  const routerMode = await page.evaluate(() => window.$docsify.routerMode);

  if (routerMode === 'history') {
    await page.evaluate(() => history.pushState(...));
  } else {
    await page.evaluate(r => (window.location.hash = r), route);
  }

  await page.waitForLoadState('networkidle');
}
```

How does that look?

Reply:
Looks good 👍

Reply:
Any reason why we can't get the `routerMode` via `page.evaluate()` instead of requiring the value to be passed as an argument to `navigateToRoute()`?

From example above:

```js
async function navigateToRoute(page, route) {
  const routerMode = await page.evaluate(() => window.$docsify.routerMode);
```

Reply:
haven't noticed that in your example :) changed